File: data-structures/hash-table.rb
####### WIP ########

class HashTable
  def initialize
    @storage = []
    @buckets = 8
    @size = 0
  end

  # verify
  def hash(str)
    hash = 5381
    str.each_byte do |char|
      p "hash #{hash}, shift #{(hash << 5)}"
      # hash = ((hash << 5) + hash) + char
      hash = (hash * 33) + char
    end
    p hash
    p @buckets
    hash * @buckets
  end


end

hash_table = HashTable.new
p hash_table.hash('test')


 # //////////////////////////////////////////////////////////
 # ///////////////  DO NOT TOUCH TEST BELOW!!!  /////////////
 # //////////////////////////////////////////////////////////

 # require 'test/unit'

 # class HashTableClassTest < Test::Unit::TestCase
 #   def test_Hash_Table_properties_existence
 #     test = Hash_Table.new()

 #     assert_respond_to(test, :buckets)
 #     assert_respond_to(test, :storage)
 #     assert_respond_to(test, :size)
 #   end

 #   def test_Hash_Table_methods_existence
 #     test = Hash_Table.new()

 #     assert_respond_to(test, :hash)
 #     assert_respond_to(test, :insert)
 #     assert_respond_to(test, :delete)
 #     assert_respond_to(test, :retrieve)
 #   end

 #   def test_Hash_Table_hashing_method
 #     test = Hash_Table.new()

 #     assert_equal(1, test.hash('hello', 8))
 #   end

 #   def test_Hash_Table_insert_a_key_value_pair
 #     test = Hash_Table.new()

 #     assert_equal(0, test.storage.length)
 #     assert_equal(0, test.size)

 #     test.insert('hello', 5)

 #     assert_equal(1, test.size)
 #     assert_equal('hello', test.storage[1][0][0])
 #     assert_equal(5, test.storage[1][0][1])
 #   end

 #   def test_Hash_Table_insert_a_second_key_value_pair
 #     test = Hash_Table.new()

 #     assert_equal(0, test.storage.length)
 #     assert_equal(0, test.size)

 #     test.insert('hello', 5)

 #     assert_equal(1, test.size)
 #     assert_equal('hello', test.storage[1][0][0])
 #     assert_equal(5, test.storage[1][0][1])

 #     test.insert('good', 10)

 #     assert_equal(2, test.size)
 #     assert_equal('good', test.storage[6][0][0])
 #     assert_equal(10, test.storage[6][0][1])
 #   end

 #   def test_Hash_Table_insert_should_handle_collisions
 #     test = Hash_Table.new()

 #     assert_equal(0, test.storage.length)
 #     assert_equal(0, test.size)

 #     test.insert('back', 5)

 #     assert_equal(1, test.size)
 #     assert_equal('back', test.storage[6][0][0])
 #     assert_equal(5, test.storage[6][0][1])

 #     test.insert('good', 10)

 #     assert_equal(2, test.size)
 #     assert_equal('good', test.storage[6][1][0])
 #     assert_equal(10, test.storage[6][1][1])
 #   end

 #   def test_Hash_Table_delete_method_should_delete_pair
 #     test = Hash_Table.new()

 #     assert_equal(0, test.storage.length)
 #     assert_equal(0, test.size)

 #     test.insert('hello', 5)

 #     assert_equal(1, test.size)
 #     assert_equal('hello', test.storage[1][0][0])
 #     assert_equal(5, test.storage[1][0][1])

 #     test.delete('hello')

 #     assert_equal(0, test.size)
 #     assert_equal(undef, test.storage[6][0][0])
 #     assert_equal(undef, test.storage[6][0][1])
 #   end

 #   def test_Hash_Table_delete_should_not_modify_size_when_key_not_found
 #     test = Hash_Table.new()

 #     test.insert('hello', 5)
 #     test.insert('good', 10)

 #     assert_equal(2, test.size)

 #     test.delete('great')

 #     expect(test.size).to.equal(2)
 #   end

 #   def test_Hash_Table_retrieve_should_return_value_when_exists
 #     test = Hash_Table.new()

 #     test.insert('hello', 5)

 #     assert_equal(5, test.retrieve('hello'))
 #   end

 #   def test_Hash_Table_retrieve_should_return_nil_when_key_does_not_exist
 #     test = Hash_Table.new()

 #     test.insert('hello', 5)

 #     assert_equal(nil, test.retrieve('good'))
 #   end

 #   def test_Hash_Table_resize_should_double_the_number_of_buckets_when_the_size_exceeds_75_percent_of_bucket_capacity
 #     test = Hash_Table.new()
 #     test.insert('hello', 5)
 #     test.insert('good', 7)
 #     test.insert('haha', 10)
 #     test.insert('blah', 2)
 #     test.insert('foo', 3)
 #     test.insert('bar', 8)
 #     test.insert('taste', 1)

 #     assert_equal(16, test.buckets)
 #   end

 #   def test_Hash_Table_resize_should_halve_the_number_of_buckets_when_the_size_drops_below_25_percent_of_bucket_capacity
 #     test = Hash_Table.new()

 #     test.insert('hello', 5)
 #     test.insert('good', 7)
 #     test.insert('haha', 10)
 #     test.insert('blah', 2)
 #     test.insert('foo', 3)
 #     test.insert('bar', 8)
 #     test.insert('taste', 1)

 #     assert_equal(16, test.buckets)

 #     test.delete('hello')
 #     test.delete('good')
 #     test.delete('haha')
 #     test.delete('blah')

 #     assert_equal(8, test.buckets)
 #   end
 # end

File: data-structures/hash-table.js

var HashTable = function() {
  this.storage = [];
  this.buckets = 8;
  this.size = 0;
}

HashTable.prototype.hash = function(str) {
  var hash = 5381;
  for (i = 0; i < str.length; i++) {
    char = str.charCodeAt(i);
    console.log('hash ', hash, 'shift ', (hash << 5))
    hash = ((hash << 5) + hash) + char; /* hash * 33 + c */
  }
  console.log('hash',hash,'buckets', this.buckets)
  return hash % this.buckets;
}

// could implement linked list here instead of array for O(1)
HashTable.prototype.insert = function(key, value) {
  var index = this.hash(key);

  if (this.storage[index] === undefined) {
    this.storage[index] = [];
    this.storage[index].push([key, value]);
    this.size++;
    this.resize();
  } else {
    var bucket = this.storage[index];

    for (var i = 0; i < bucket.length; i++) {
      if (bucket[i][0] == key) {
        bucket[i][1] = value;
        return;
      }
    }

    bucket.push([key, value]);
    this.size++;
    this.resize();
  }
}

HashTable.prototype.delete = function(key) {
  var index = this.hash(key);

  if (this.storage[index] === undefined) {
    console.log('key not found');
    return;
  } else {
    var bucket = this.storage[index]

    for (var i=0; i<bucket.length; i++) {
      if (bucket[i][0] === key) {
        var temp = bucket[i][1];
        bucket.splice(i,1);
        this.size--;
        this.resize();
        return temp;
      }
    }
    console.log('key not found');
  }
}

HashTable.prototype.retrieve = function(key) {
  var index = this.hash(key);

  if (this.storage[index] === undefined) {
    console.log('key not found');
    return;
  } else {
    var bucket = this.storage[index]

    for (var i=0; i<bucket.length; i++) {
      if (bucket[i][0] === key) {
        return bucket[i][1];
      }
    }
    console.log('key not found');
  }
}

HashTable.prototype.resize = function() {
  var allElements = [];

  if (this.size > (0.75 * this.buckets)) {
    this.buckets *= 2;
    this.storage.forEach(function(bucket) {
      if (bucket !== undefined) {
        bucket.forEach(function(tuple) {
          allElements.push(tuple);
        })
      }
    })
    this.storage = []
    this.size = 0;
    var that = this;
    allElements.forEach(function(tuple) {
      that.insert(tuple[0], tuple[1])
    })

    console.log('HashTable has been doubled in size');
  } else if (this.buckets > 8 && this.size < (0.25 * this.buckets)) {
    this.buckets *= 0.5;
    this.storage.forEach(function(bucket) {
      if (bucket !== undefined) {
        bucket.forEach(function(tuple) {
          allElements.push(tuple);
        })
      }
    })
    this.storage = []
    this.size = 0;
    var that = this;
    allElements.forEach(function(tuple) {
      that.insert(tuple[0], tuple[1])
    })

    console.log('HashTable has been halved in size');
  }
}

hashTable = new HashTable
console.log(hashTable.hash('test'))

////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////

// var expect = require('chai').expect;

// describe('hash table class ', function(){

//   describe('HashTable properties', function(){
//     it('should have properties storage, buckets, and size', function(){
//       var test = new HashTable();

//       expect(test).to.have.property('storage');
//       expect(test).to.have.property('buckets');
//       expect(test).to.have.property('size');
//       expect(test.storage.length).to.equal(0);
//       expect(test.size).to.equal(0);
//     });
//   });

//   describe('HashTable methods existence', function(){
//     it('should have methods hash, insert, delete, and retrieve', function(){
//       var test = new HashTable();

//       expect(test).to.respondTo('hash');
//       expect(test).to.respondTo('insert');
//       expect(test).to.respondTo('delete');
//       expect(test).to.respondTo('retrieve');
//     });
//   });

//   describe('HashTable hash method', function(){
//     it('should return an index from an inputted string', function(){
//       var test = new HashTable();

//       var expected = test.hash('hello');
//       expect(expected).to.equal(1);
//     })
//   });

//   describe('HashTable insert method', function(){
//     it('should be able to insert a key-value pair', function(){
//       var test = new HashTable();

//       expect(test.storage.length).to.equal(0);
//       expect(test.size).to.equal(0);

//       test.insert('hello', 5);

//       expect(test.size).to.equal(1);
//       expect(test.storage[1][0][0]).to.equal('hello');
//       expect(test.storage[1][0][1]).to.equal(5);
//     });

//     it('should be able to insert a second key-value pair', function(){
//       var test = new HashTable();

//       expect(test.storage.length).to.equal(0);
//       expect(test.size).to.equal(0);

//       test.insert('hello', 5);

//       expect(test.size).to.equal(1);
//       expect(test.storage[1][0][0]).to.equal('hello');
//       expect(test.storage[1][0][1]).to.equal(5);

//       test.insert('good', 10);

//       expect(test.size).to.equal(2);
//       expect(test.storage[6][0][0]).to.equal('good');
//       expect(test.storage[6][0][1]).to.equal(10);
//     });

//     it('should be able to handle collisions', function(){
//       var test = new HashTable();

//       expect(test.storage.length).to.equal(0);
//       expect(test.size).to.equal(0);

//       test.insert('good', 5);

//       expect(test.size).to.equal(1);
//       expect(test.storage[6][0][0]).to.equal('good');
//       expect(test.storage[6][0][1]).to.equal(5);

//       test.insert('back', 10);

//       expect(test.size).to.equal(2);
//       expect(test.storage[6][1][0]).to.equal('back');
//       expect(test.storage[6][1][1]).to.equal(10);
//     });
//   });

//   describe('HashTable delete method', function(){
//     it('should delete a key-value pair', function(){
//       var test = new HashTable();

//       expect(test.storage.length).to.equal(0);
//       expect(test.size).to.equal(0);

//       test.insert('hello', 5);

//       expect(test.size).to.equal(1);
//       expect(test.storage[1][0][0]).to.equal('hello');
//       expect(test.storage[1][0][1]).to.equal(5);

//       test.delete('hello');

//       expect(test.size).to.equal(0);
//       expect(test.storage[1][0]).to.equal(undefined);
//       expect(test.storage[1][0]).to.equal(undefined);

//     });

//     it('should not modify the size when deleting a key-value pair that does not exist', function(){
//       var test = new HashTable();
//       test.insert('hello', 5);
//       test.insert('good', 10);

//       expect(test.size).to.equal(2);

//       test.delete('great');

//       expect(test.size).to.equal(2);
//     });
//   });

//   describe('HashTable retrieve method', function(){
//     it('should return true for a key-value pair that exists', function(){
//       var test = new HashTable();

//       test.insert('hello', 5);

//       expect(test.retrieve('hello')).to.equal(5);


//     });

//     it('should return false for a key-value pair that does not exist', function(){
//       var test = new HashTable();

//       test.insert('hello', 5);

//       expect(test.retrieve('good')).to.equal(null);
//     });
//   });

//   describe('HashTable resize method', function(){
//     it('should double the number of buckets when the size exceeds 75% of buckets capacity', function(){
//       var test = new HashTable();

//       test.insert('hello', 5);
//       test.insert('good', 7);
//       test.insert('haha', 10);
//       test.insert('blah', 2);
//       test.insert('foo', 3);
//       test.insert('bar', 8);
//       test.insert('taste', 1);

//       expect(test.buckets).to.equal(16);


//     });

//     it('should halve the number of buckets when the size drops below 25% of bucket capacity', function(){
//       var test = new HashTable();

//       test.insert('hello', 5);
//       test.insert('good', 7);
//       test.insert('haha', 10);
//       test.insert('blah', 2);
//       test.insert('foo', 3);
//       test.insert('bar', 8);
//       test.insert('taste', 1);

//       expect(test.buckets).to.equal(16);

//       test.delete('hello');
//       test.delete('good');
//       test.delete('haha');
//       test.delete('blah');

//       expect(test.buckets).to.equal(8);

//     });
//   });

// })
